rclc_parameter/src/rclc_parameter/parameter_server.c

```diff
@@ -827,6 +1545,134 @@ rclc_parameter_server_init_service(
+  rclc_parameter_server_t * parameter_server,
+  const char * parameter_name, bool read_only)
+{
+  if (parameter_server->on_callback) {
+    return RCLC_PARAMETER_DISABLED_ON_CALLBACK;
+  }
+
+  size_t index = rclc_parameter_search_index(&parameter_server->parameter_list, parameter_name);
+
+  if (index >= parameter_server->parameter_list.size) {
+    return RCL_RET_ERROR;
+  }
+
+  ParameterDescriptor * parameter_descriptor = &parameter_server->parameter_descriptors.data[index];
+
+  // Set flag
+  parameter_descriptor->read_only = read_only;
+
+  return RCL_RET_OK;
+}
+
+rcl_ret_t rclc_add_parameter_constraint_double(
+  rclc_parameter_server_t * parameter_server,
+  const char * parameter_name, double from_value,
+  double to_value, double step)
+{
+  if (parameter_server->on_callback) {
+    return RCLC_PARAMETER_DISABLED_ON_CALLBACK;
+  }
+
+  size_t index = rclc_parameter_search_index(&parameter_server->parameter_list, parameter_name);
+
+  if (index >= parameter_server->parameter_list.size) {
+    return RCL_RET_ERROR;
+  }
+
+  ParameterDescriptor * parameter_descriptor = &parameter_server->parameter_descriptors.data[index];
+
+  if (parameter_descriptor->type != RCLC_PARAMETER_DOUBLE) {
+    return RCL_RET_ERROR;
+  }
+
+  parameter_descriptor->floating_point_range.data->from_value = from_value;
+  parameter_descriptor->floating_point_range.data->to_value = to_value;
+  parameter_descriptor->floating_point_range.data->step = step;
+  parameter_descriptor->floating_point_range.size = 1;
+
+  return RCL_RET_OK;
+}
+
+rcl_ret_t rclc_add_parameter_constraint_integer(
+  rclc_parameter_server_t * parameter_server,
+  const char * parameter_name, int64_t from_value,
+  int64_t to_value, uint64_t step)
+{
+  if (parameter_server->on_callback) {
+    return RCLC_PARAMETER_DISABLED_ON_CALLBACK;
+  }
+
+  size_t index = rclc_parameter_search_index(&parameter_server->parameter_list, parameter_name);
+
+  if (index >= parameter_server->parameter_list.size) {
+    return RCL_RET_ERROR;
+  }
+
+  ParameterDescriptor * parameter_descriptor = &parameter_server->parameter_descriptors.data[index];
+
+  if (parameter_descriptor->type != RCLC_PARAMETER_INT) {
+    return RCL_RET_ERROR;
+  }
+
+  parameter_descriptor->integer_range.data->from_value = from_value;
+  parameter_descriptor->integer_range.data->to_value = to_value;
+  parameter_descriptor->integer_range.data->step = step;
+  parameter_descriptor->integer_range.size = 1;
+
+  return RCL_RET_OK;
+}
+
+rcl_ret_t rclc_parameter_execute_callback(
+  rclc_parameter_server_t * parameter_server,
+  const Parameter * old_param, const Parameter * new_param)
+{
+  bool ret = true;
+
+  if (parameter_server->on_modification) {
+    parameter_server->on_callback = true;
+    ret = parameter_server->on_modification(old_param, new_param, parameter_server->context);
+    parameter_server->on_callback = false;
+  }
+
+  return ret ? RCL_RET_OK : RCLC_PARAMETER_MODIFICATION_REJECTED;
```

Comment:
sure? RCL_RET_OK = 0
see:
https://docs.ros2.org/eloquent/api/rcl/types_8h.html
https://docs.ros2.org/eloquent/api/rmw/ret__types_8h.html#a1fd606c616b0bfb93186c9bae623832a

if ret== `RCL_RET_OK` then you are returning `RCLC_PARAMETER_MODIFICATION_REJECTED`




for readability, I would write 
`return (ret==RCL_RET_OK) ? RCL_RET_OK : RCLC_PARAMETER_MODIFICATION_REJECTED;`

or what is the return value of `on_modification` function?

Reply:
`ret` returned from user `on_modification` callback, which returns a bool value to accept or reject modifications:
https://github.com/ros2/rclc/blob/eb4b3ca091928a750add48eae0ba0c98e7c06e43/rclc_parameter/include/rclc_parameter/rclc_parameter.h#L98-L101

I understand that naming this boolean var `ret` can be confusing.